torch/utils/data/_utils/worker.py

```diff
@@ -75,8 +123,37 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
 
         data_queue.cancel_join_thread()
 
-        if init_fn is not None:
-            init_fn(worker_id)
+        global _worker_info
+        _worker_info = WorkerInfo(id=worker_id, num_workers=num_workers,
+                                  seed=seed, dataset=dataset)
+
+        from torch.utils.data import _DataLoaderStrategy
+
+        init_exception = None
+
+        try:
+            if init_fn is not None:
+                init_fn(worker_id)
+
+            if mode == _DataLoaderStrategy.Iterable:
+                dataset_iter = iter(dataset)
+
+        except Exception:
+            init_exception = ExceptionWrapper(sys.exc_info())
+
+        # When using Iterable mode, some worker can exit earlier than others due
+        # to the IterableDataset behaving differently for different workers.
+        # When such things happen, an `IterableDatasetStopIteration` object is
+        # sent over to the main process with the ID of this worker, so that the
+        # main process won't send more tasks to this worker, and will send
+        # `None` to this worker to properly exit it.
```

Comment:
Why do we even bother with sending anything from master to workers in iterable mode? Because you want to bound the number of batches awaiting in memory? Couldn't we simply use an atomic IPC counter for that purpose?

Reply:
> Because you want to bound the number of batches awaiting in memory?

Yep, that is the main reason. I do agree that an simple atomic IPC would probably suffice the majority of the work. But it would require adding more than just that if we do not use `queue`s at all. Consider the following requirements, all of which we spent significant amount of work to make sure work with `queue` communication:

1. We want data loading order to be deterministic (conditioned on seeds). 

    Following the current DataLoader design, we index the workloads to properly track which is the next batch to return. This allows the worker to receive an `index` and send it back. Admittedly this could also be done if we just for each worker keep a FIFO queue of indices of all work loads, but the existing code does so naturally with `queues`. 

3. Signaling the workers to quit. 

    Currently we use sending `None` to workers as the authority signal to terminate. Yes, we could use another `mp.Event` to do so (`done_event` is not exactly for this purpose), but the existing code does so with `queue`s in a quite robust way.

2. Do not hang.

   We took sooo long to build all these mechanisms for `queue` communication (hopefully) correctly to prevent hanging. Implementing new ways to signal workers using semaphores or `mp.Condition`s could be tricky (although I admit I didn't spend too much time thinking about the details).

Reinventing these wheels would be a lot of work and might be hard to get right. So I just went with the `queue`s, which also nicely reduce code duplication and are probably implemented under the hood as semaphores and pipes anyways.

Reply:
Ok, this makes sense. I just wanted to clarify that. Thanks for the detailed respose!